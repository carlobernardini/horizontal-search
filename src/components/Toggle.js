import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import uniqueId from 'lodash/uniqueId';
import Caption from './Caption';
import FilterInfo from './FilterInfo';
import { DISPLAY_NAME_TOGGLE, NAME_PREFIX_TOGGLE } from '../constants';

const Toggle = ({ disabled, information, label, onUpdate, option, toggleStateLabel, value }) => {
    const id = uniqueId(NAME_PREFIX_TOGGLE);

    return (
        <div
            className="orizzonte__filter"
        >
            { !!label && (
                <FilterInfo
                    information={ information }
                />
            ) }
            <Caption>
                { label }
            </Caption>
            <div
                className={ classNames('orizzonte__toggle', {
                    'orizzonte__toggle--disabled': disabled
                }) }
            >
                <input
                    checked={ !!value }
                    className="orizzonte__toggle-input"
                    disabled={ disabled }
                    id={ id }
                    type="checkbox"
                    value={ option.value }
                    onChange={ (e) => {
                        const { checked } = e.target;
                        onUpdate(checked ? option.value : null);
                    }}
                />
                <label
                    htmlFor={ id }
                    className="orizzonte__toggle-label"
                >
                    <span
                        className="orizzonte__toggle-text"
                    >
                        { option.label || option.value }
                    </span>
                    { !!toggleStateLabel && (
                        <span
                            className={ classNames('orizzonte__toggle-state', {
                                'orizzonte__toggle-state--on': !!value
                            }) }
                        >
                            { value ? toggleStateLabel.on : toggleStateLabel.off }
                        </span>
                    ) }
                    <span
                        className="orizzonte__toggle-control"
                    />
                </label>
            </div>
        </div>
    );
};

Toggle.displayName = DISPLAY_NAME_TOGGLE;

Toggle.propTypes = {
    /** Disables the toggle switch */
    disabled: PropTypes.bool,
    /** Information tooltip for this filter (requires label prop to be defined */
    information: PropTypes.string,
    /** Label for this filter section */
    label: PropTypes.string,
    /** Internal callback for when filter state has changed */
    onUpdate: PropTypes.func,
    /** The label and value for this field. Value is used when switched on. */
    option: PropTypes.shape({
        label: PropTypes.string,
        value: PropTypes.oneOfType([
            PropTypes.number,
            PropTypes.string
        ]).isRequired
    }).isRequired,
    /** String indicators for the current toggle state, to be shown next to
        the toggle switch button */
    toggleStateLabel: PropTypes.shape({
        on: PropTypes.string.isRequired,
        off: PropTypes.string.isRequired
    }),
    value: PropTypes.oneOfType([
        PropTypes.number,
        PropTypes.string
    ])
};

Toggle.defaultProps = {
    disabled: false,
    information: null,
    label: null,
    onUpdate: () => {},
    toggleStateLabel: null,
    value: null
};

export default Toggle;
