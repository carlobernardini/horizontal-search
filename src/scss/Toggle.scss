@import 'base';

.orizzonte__toggle {
    &-input {
        display: none;

        &:checked {
            + .orizzonte__toggle-label {
                .orizzonte__toggle-control {
                    background: $color-d;

                    &:before {
                        left: calc(100% - 20px);
                        transform: scale(2);
                        opacity: 0;
                        transition: all .4s ease;
                    }
                    &:after {
                        left: calc(100% - 20px);
                    }
                }
            }
        }
    }
    &-label {
        display: flex;
        justify-content: space-between;
        align-items: center;
        cursor: pointer;
    }
    &-text {
        padding: 5px 0;
        color: $text-color;
        flex-grow: 2;
    }
    &-control {
        display: inline-block;
        vertical-align: middle;
        cursor: pointer;
        text-indent: -9999px;
        width: 38px;
        height: 22px;
        background: darken($gray-lighter, 5%);
        box-shadow: inset 0 1px 0 rgba(0,0,0, .3);
        border-radius: 100px;
        position: relative;
        transition: all .5s;

        &:before {
            content: '';
            top: 2px;
            left: 2px;
            position: absolute;
            width: 18px;
            height: 18px;
            background: $color-a;
            display: block;
            transform: scale(0);
            opacity: 1;
            border-radius: 50%;
        }
        &:after {
            content: '';
            position: absolute;
            top: 2px;
            left: 2px;
            width: 18px;
            height: 18px;
            background: #fff;
            border-radius: 50%;
            transition: all .5s;
            box-shadow: inset 0 -1px 0 rgba(0,0,0, .5);
        }
    }
    &-state {
        color: $text-color-muted;
        text-transform: uppercase;
        font-size: .75em;
        margin: 0 10px;
        transition: color .3s;
        white-space: nowrap;
        text-overflow: ellipsis;

        &--on {
            color: $color-a;
        }
    }
    &--disabled {
        .orizzonte__toggle-label,
        .orizzonte__toggle-control {
            cursor: not-allowed;
        }

        .orizzonte__toggle-text {
            color: $text-color-muted;
        }
    }
}