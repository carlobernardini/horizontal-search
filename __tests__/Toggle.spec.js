import React from 'react';
import Toggle from '../src/components/Toggle';

describe('<Toggle />', () => {
    it('should render a toggle filter', () => {
        const onUpdate = jest.fn();

        const wrapper = shallow(
            <Toggle
                label="Some toggle"
                option={{
                    label: 'Toggle me on or off',
                    value: 'on'
                }}
                onUpdate={ onUpdate }
            />
        );

        expect(wrapper).toMatchSnapshot();
        wrapper.find('.orizzonte__toggle-input').simulate('change', {
            target: {
                checked: true
            }
        });
        expect(onUpdate).toHaveBeenCalledWith('on');

        wrapper.setProps({
            value: 'on'
        });
        wrapper.find('.orizzonte__toggle-input').simulate('change', {
            target: {
                checked: false
            }
        });
        expect(onUpdate).toHaveBeenCalledWith(null);
    });

    it('should render a toggle filter that only has a value property (no label)', () => {
        const wrapper = shallow(
            <Toggle
                label="Some toggle"
                option={{
                    value: 'Active'
                }}
            />
        );

        expect(wrapper).toMatchSnapshot();
    });

    it('should render a disabled toggle filter', () => {
        const wrapper = shallow(
            <Toggle
                label="Some toggle"
                option={{
                    label: 'Toggle me on or off',
                    value: 'on'
                }}
                disabled
            />
        );

        expect(wrapper).toMatchSnapshot();
    });

    it('should render a toggle filter with toggle state indicators', () => {
        const wrapper = shallow(
            <Toggle
                label="Some toggle"
                option={{
                    label: 'Toggle me on or off',
                    value: 'on'
                }}
                toggleStateLabel={{
                    on: 'Active',
                    off: 'Inactive'
                }}
            />
        );

        expect(wrapper).toMatchSnapshot();

        wrapper.setProps({
            value: 'on'
        });
        expect(wrapper).toMatchSnapshot();        
    });
});
