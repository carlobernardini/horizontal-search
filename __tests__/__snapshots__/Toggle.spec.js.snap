// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Toggle /> should render a disabled toggle filter 1`] = `
<div
  className="orizzonte__filter"
>
  <OrizzonteFilterInfo
    information={null}
  />
  <OrizzonteFilterCaption>
    Some toggle
  </OrizzonteFilterCaption>
  <div
    className="orizzonte__toggle orizzonte__toggle--disabled"
  >
    <input
      checked={false}
      className="orizzonte__toggle-input"
      disabled={true}
      id="toggle-4"
      onChange={[Function]}
      type="checkbox"
      value="on"
    />
    <label
      className="orizzonte__toggle-label"
      htmlFor="toggle-4"
    >
      <span
        className="orizzonte__toggle-text"
      >
        Toggle me on or off
      </span>
      <span
        className="orizzonte__toggle-control"
      />
    </label>
  </div>
</div>
`;

exports[`<Toggle /> should render a toggle filter 1`] = `
<div
  className="orizzonte__filter"
>
  <OrizzonteFilterInfo
    information={null}
  />
  <OrizzonteFilterCaption>
    Some toggle
  </OrizzonteFilterCaption>
  <div
    className="orizzonte__toggle"
  >
    <input
      checked={false}
      className="orizzonte__toggle-input"
      disabled={false}
      id="toggle-1"
      onChange={[Function]}
      type="checkbox"
      value="on"
    />
    <label
      className="orizzonte__toggle-label"
      htmlFor="toggle-1"
    >
      <span
        className="orizzonte__toggle-text"
      >
        Toggle me on or off
      </span>
      <span
        className="orizzonte__toggle-control"
      />
    </label>
  </div>
</div>
`;

exports[`<Toggle /> should render a toggle filter that only has a value property (no label) 1`] = `
<div
  className="orizzonte__filter"
>
  <OrizzonteFilterInfo
    information={null}
  />
  <OrizzonteFilterCaption>
    Some toggle
  </OrizzonteFilterCaption>
  <div
    className="orizzonte__toggle"
  >
    <input
      checked={false}
      className="orizzonte__toggle-input"
      disabled={false}
      id="toggle-3"
      onChange={[Function]}
      type="checkbox"
      value="Active"
    />
    <label
      className="orizzonte__toggle-label"
      htmlFor="toggle-3"
    >
      <span
        className="orizzonte__toggle-text"
      >
        Active
      </span>
      <span
        className="orizzonte__toggle-control"
      />
    </label>
  </div>
</div>
`;

exports[`<Toggle /> should render a toggle filter with toggle state indicators 1`] = `
<div
  className="orizzonte__filter"
>
  <OrizzonteFilterInfo
    information={null}
  />
  <OrizzonteFilterCaption>
    Some toggle
  </OrizzonteFilterCaption>
  <div
    className="orizzonte__toggle"
  >
    <input
      checked={false}
      className="orizzonte__toggle-input"
      disabled={false}
      id="toggle-5"
      onChange={[Function]}
      type="checkbox"
      value="on"
    />
    <label
      className="orizzonte__toggle-label"
      htmlFor="toggle-5"
    >
      <span
        className="orizzonte__toggle-text"
      >
        Toggle me on or off
      </span>
      <span
        className="orizzonte__toggle-state"
      >
        Inactive
      </span>
      <span
        className="orizzonte__toggle-control"
      />
    </label>
  </div>
</div>
`;

exports[`<Toggle /> should render a toggle filter with toggle state indicators 2`] = `
<div
  className="orizzonte__filter"
>
  <OrizzonteFilterInfo
    information={null}
  />
  <OrizzonteFilterCaption>
    Some toggle
  </OrizzonteFilterCaption>
  <div
    className="orizzonte__toggle"
  >
    <input
      checked={true}
      className="orizzonte__toggle-input"
      disabled={false}
      id="toggle-6"
      onChange={[Function]}
      type="checkbox"
      value="on"
    />
    <label
      className="orizzonte__toggle-label"
      htmlFor="toggle-6"
    >
      <span
        className="orizzonte__toggle-text"
      >
        Toggle me on or off
      </span>
      <span
        className="orizzonte__toggle-state orizzonte__toggle-state--on"
      >
        Active
      </span>
      <span
        className="orizzonte__toggle-control"
      />
    </label>
  </div>
</div>
`;
